<div style="border:solid green 2px; padding: 20px">

<b>Привет, Артем!</b>

Меня зовут Алина Соловьева и я буду проверять твою работу. Попробую помочь сделать тебе её еще лучше! В Практикуме принято общение на ты, но если это дискомфортно, то сообщи об этом)

Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>☁️Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b>⛅️ Так выделены небольшие замечания.Такие замечания в проекте допустимы, это пространство для улучшений в будущем.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Так я выделяю успешные идеи.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:

<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>    

Всё это поможет выполнить повторную проверку твоего проекта оперативнее.

# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**
Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Молодец, что начинаешь с введения - это важная часть проекта.</div>

Подключаемся к базе данных

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

выводим первые 5 строк таблицы books

In [4]:
query = '''SELECT * FROM books LIMIT 5'''

In [5]:
con=engine.connect()

In [6]:
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

выводим первые 5 строк таблицы author

In [7]:
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

выводим первые 5 строк таблицы publishers

In [8]:
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

выводим первые 5 строк таблицы ratings

In [9]:
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

выводим первые 5 строк таблицы reviews

In [10]:
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Хорошо. Познакомились с данными, посмотрели содержимое таблиц и заодно проверили соединение с сервером.</div>

задание 1  

- Посчитайте, сколько книг вышло после 1 января 2000 года;

In [11]:
query = '''SELECT count(book_id)
            FROM books 
            where cast(publication_date as date) > '2000-01-01' '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 года вышло 819 книг


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Отлично.</div>

задание 2  
Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [12]:
query = '''SELECT b.book_id,
                    b.title,
                    avg(rating) as avg_rating,
                    count(distinct review_id) as cnt_review
            FROM books as b
            left join ratings as r on b.book_id=r.book_id
            left join reviews as re on b.book_id = re.book_id
            group by b.book_id
            order by cnt_review desc
                        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  avg_rating  \
0        948                            Twilight (Twilight  #1)    3.662500   
1        963                                Water for Elephants    3.977273   
2        734                                   The Glass Castle    4.206897   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
4        695  The Curious Incident of the Dog in the Night-Time    4.081081   
..       ...                                                ...         ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel    3.666667   
996      808                            The Natural Way to Draw    3.000000   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites    5.000000   
998      221                          Essential Tales and Poems    4.000000   
999      191  Disney's Beauty and the Beast (A Little Golden...    4.000000   

     cnt_review  
0             7  
1             6  
2             6  
3             6  
4             6  
..          ...  
995           0  
996           0  
997           0  
998           0  
999           0  

[1000 rows x 4 columns]

Получили таблицу из 1000 строк, т.е. из 1000 книг, в которых указан id книги,название книги, средняя оценка и колличество обзоров


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Да, это верно.</div>

задание 3  
Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [13]:
query = '''
select
        p.publisher,
        count(b.book_id)
from books as b
left join publishers as p on b.publisher_id=p.publisher_id
where b.num_pages > 50
group by p.publisher_id
order by count(b.book_id) desc
limit 1
                        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count
0  Penguin Books     42

Penguin Books выпустило 42 книги, что является наибольшим показателем среди всех издательств


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Да, это так.</div>

задание 4 
Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [14]:
query = '''
with t as (
select 
    b.book_id,
    b.author_id,
    count(r.rating_id) as cnt_rating,
    avg(r.rating) as avg_rating
from books as b
join ratings as r on b.book_id=r.book_id
group by b.book_id
having count(r.rating_id) >= 50 
order by avg_rating desc)

select 
    author,
    avg(avg_rating)
from authors as a
right join t on a.author_id=t.author_id
group by author
order by avg(avg_rating) desc
limit 1
                        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

Наивысшая средняя оценка всех книг автора составляет 4,28 у J.K.Rowling 
(в расчет принимались только те книги у которых более 50 оценок)


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Самое сложное задание из выборки выполнено с первой попытки!</div>

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [15]:
query = '''
with t as (
SELECT 
        username,
        count(r.rating_id) as cnt
from books as b
join ratings as r on b.book_id=r.book_id
group by r.username
having count(r.rating_id) >48
)


select 
        count(review_id)/count(distinct(username)) as avg_review
from reviews
where username in (select username from t)

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg_review
0          24

Пользователи, которые ставят более 48 оценков, в среднем пишут по 24 отзыва


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Точно!</div>


<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>Нужно только добавить финальный вывод и все будет принято :) Других критических замечаний нет. </div>

вывод: было загружено 5 таблиц и просмотренных первые 5 строк каждой таблицы.  
далее было выполнено первое задание и обнаружено, что после первого января 2000 годабыло выпущено 819 книг  
выполнено второе задание , в котором была подсчитано средняя оценка и количество обзоров каждой книги
выполнено третье задание, в котором было обнаружено издательство, напечатавшее наибольшее количество книг. это  penguin Books  
выполнено четвёртое задание, в котором была обнаружен автор с наибольшей средней оценкой книг, среди книг с более чем 50 оценками  
пятое задание, в котором было установлено что пользователи , которые ставят более 48 оценок, в среднем пишут по 24 отзыва

все дания были выполнены с использованием sql




<div class="alert alert-success">
<b>Комментарий ревьюера v2.0:</b>☀️Проект принят, поздравляю тебя ;) Успехов на поприще аналитики!</div>